src/go/util/httppattern/uri_template.go

```diff
@@ -0,0 +1,411 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httppattern
+
+import (
+	"bytes"
+)
+
+// UriTemplate is used to syntax pairse uri templates. It is based on the grammar
+// on https://github.com/googleapis/googleapis/blob/e5211c547d63632963f9125e2b333185d57ff8f6/google/api/http.proto#L224.
+type UriTemplate struct {
+	Segments  []string
+	Verb      string
+	Variables []*variable
+}
+
+// The info about a variable binding {variable=subpath} in the template.
+type variable struct {
+	// Specifies the range of segments [start_segment, end_segment) the
+	// variable binds to. Both start_segment and end_segment are 0 based.
+	// end_segment can also be negative, which means that the position is
+	// specified relative to the end such that -1 corresponds to the end
+	// of the path.
+	StartSegment int
+	EndSegment   int
+
+	// The path of the protobuf field the variable binds to.
+	FieldPath []string
+
+	// Do we have a ** in the variable template?
+	HasWildcardPath bool
+}
+
+// HTTP Template Grammar:
+// Questions:
+//
+// Template = "/" | "/" Segments [ Verb ] ;
+// Segments = Segment { "/" Segment } ;
+// Segment  = "*" | "**" | LITERAL | Variable ;
+// Variable = "{" FieldPath [ "=" Segments ] "}" ;
+// FieldPath = IDENT { "." IDENT } ;
+// Verb     = ":" LITERAL ;
+type parser struct {
+	input      string
+	tb         int
+	te         int
+	inVariable bool
+	segments   []string
+	verb       string
+	variables  []*variable
+}
+
+func Parse(ht string) *UriTemplate {
+	if ht == "/" {
+		return &UriTemplate{}
+	}
+
+	p := parser{
+		input: ht,
+	}
+	if !p.parse() || !p.validateParts() {
+		return nil
```

Comment:
We need some form of error reporting. At the very least, we can just return an error:

```
return nil, fmt.Errorf("Invalid http template (%v)", ht)
```

Reply:
The error info not under parsing environment will be meaningless and trivial. For example, 
- under such parsing stage, missing some components
- don't meet some expected part.

As for the `Invalid http template (%v)`, caller can add too if it is not successfully parsed.